### PR TITLE
Test shaded UGI

### PR DESCRIPTION
**Pull request by wendigo**
## Release notes

( ) This is not user-visible or is docs only, and no release notes are required.
( ) Release notes are required. Please propose a release note for me.
( ) Release notes are required, with the following suggested text:

```markdown
# Section
* Fix some things. ({issue}`issuenumber`)
```

### CODE REVIEW

**Comment by wendigo** [comment]
Supersedes https://github.com/trinodb/trino/pull/21798

File: lib/trino-hdfs/src/main/java/io/trino/hdfs/authentication/UserGroupInformationUtils.java [hunk 1]
```diff
@@ -24,14 +25,19 @@ private UserGroupInformationUtils() {}
...
+                    return new ResultOrException<>(action.run(), null);
+                }
+                catch (Throwable e) {
+                    return new ResultOrException<>(null, e);
+                }
+            }).get();
+        }
+        catch (IOException | InterruptedException e) {
+            throw new RuntimeException(e);
+        }
```
Comment (electrum) [inline lib/trino-hdfs/src/main/java/io/trino/hdfs/authentication/UserGroupInformationUtils.java:40]:
```suggestion
        catch (IOException e) {
            throw new UncheckedIOException(e);
        }
        catch (InterruptedException e) {
            Thread.currentThread().interrupt();
            throw new RuntimeException(e);
        }
```

**Comment by electrum** [comment]
The exception wrapping behavior is broken:

```
Error:  io.trino.hdfs.TestFileSystemCache.testFileSystemCacheException -- Time elapsed: 0.093 s <<< FAILURE!
java.lang.AssertionError: 

Expecting actual throwable to be an instance of:
  java.io.IOException
but was:
  java.io.UncheckedIOException: java.io.IOException: FileSystem max cache size has been reached: 1000
	at io.trino.hdfs.authentication.UserGroupInformationUtils.executeActionInDoAs(UserGroupInformationUtils.java:39)
	at io.trino.hdfs.authentication.ImpersonatingHdfsAuthentication.doAs(ImpersonatingHdfsAuthentication.java:41)
	at io.trino.hdfs.HdfsEnvironment.getFileSystem(HdfsEnvironment.java:109)
	...(64 remaining lines not displayed - this can be changed with Assertions.setMaxStackTraceElementsDisplayed)
	at io.trino.hdfs.TestFileSystemCache.testFileSystemCacheException(TestFileSystemCache.java:101)
	at java.base/java.lang.reflect.Method.invoke(Method.java:5[80](https://github.com/trinodb/trino/actions/runs/8943657564/job/24613543887?pr=21814#step:6:81))
	at java.base/java.util.concurrent.RecursiveAction.exec(RecursiveAction.java:194)
	at java.base/java.util.concurrent.ForkJoinTask.doExec(ForkJoinTask.java:507)
	at java.base/java.util.concurrent.ForkJoinPool$WorkQueue.topLevelExec(ForkJoinPool.java:1491)
	at java.base/java.util.concurrent.ForkJoinPool.scan(ForkJoinPool.java:2073)

**Comment by electrum** [comment]
Given the new signature of `callAs()` which can throw `IOException` and `InterruptedException`, I don't see a way to make the existing exception handling work. However, I think it was overly complicated, and we can simplify it:

* Rename `GenericExceptionAction` to `IOExceptionAction` and make it `throws IOException`
* Make the code calling `callAs()` convert `InterruptedException` into `InterruptedIOException` (seems better than `RuntimeException` in this case)

I tried this locally and all of the callers of `HdfsAuthentication.doAs()` expect `IOException`, so there is no need for the generic exception handling. We can probably also remove `ResultOrException` and maybe remove `UserGroupInformationUtils` entirely.

**Comment by electrum** [comment]
I put up a new PR with these changes: https://github.com/trinodb/trino/pull/21826

**Pull request closed without merging**